plutus-ledger-api/exe/analyse-script-events/Main.hs

```diff
@@ -0,0 +1,343 @@
+{-# LANGUAGE TypeApplications #-}
+
+-- | Various analyses of events in mainnet script dumps.
+-- This only deals with PlutusV1 and PlutusV2 script events because
+-- PlutusLedgerApi.Test.EvaluationEvent (and hence the scriptdump job) doesn't
+-- know about anything else yet.
+
+module Main (main) where
+
+import LoadScriptEvents (eventsOf, loadEvents)
+
+import PlutusCore.Data as Data (Data (..))
+import PlutusCore.Default (DefaultUni (DefaultUniData), Some (..), ValueOf (..))
+import PlutusCore.Evaluation.Machine.CostStream (sumCostStream)
+import PlutusCore.Evaluation.Machine.ExMemoryUsage (ExMemoryUsage, flattenCostRose, memoryUsage)
+import PlutusCore.Pretty (display)
+import PlutusLedgerApi.Common
+import PlutusLedgerApi.Test.EvaluationEvent
+import PlutusLedgerApi.V1 qualified as V1
+import PlutusLedgerApi.V2 qualified as V2
+import PlutusTx.AssocMap qualified as M
+import UntypedPlutusCore as UPLC
+
+import Control.Lens hiding (List)
+import Control.Monad.Writer.Strict
+import Data.List (intercalate)
+import Data.SatInt (fromSatInt)
+import System.Directory.Extra (listFiles)
+import System.Environment (getArgs, getProgName)
+import System.FilePath (isExtensionOf)
+import System.IO (stderr)
+import Text.Printf (hPrintf, printf)
+
+-- | The type of a generic analysis function
+type EventAnalyser
+    =  EvaluationContext
+    -> [Integer]  -- cost parameters
+    -> ScriptEvaluationEvent
+    -> IO ()
+
+-- Analyse values in ScriptContext
+
+-- Script purpose: this is the same for V1 and V2, but changes in V3
+stringOfPurposeV1 :: V1.ScriptPurpose -> String
+stringOfPurposeV1 = \case
+    V1.Minting    _ -> "V1 Minting"      -- Script arguments are [redeemer, context]
+    V1.Spending   _ -> "V1 Spending"     -- Script arguments are [datum, redeemer, context]
+    V1.Rewarding  _ -> "V1 Rewarding"    -- Script arguments appear to be [redeemer, context]
+    V1.Certifying _ -> "V1 Certifying"   -- Script arguments appear to be [redeemer, context]
+
+stringOfPurposeV2 :: V2.ScriptPurpose -> String
+stringOfPurposeV2 = \case
+    V2.Minting    _ -> "V2 Minting"
+    V2.Spending   _ -> "V2 Spending"
+    V2.Rewarding  _ -> "V2 Rewarding"
+    V2.Certifying _ -> "V2 Certifying"
+
+shapeOfValue :: V1.Value  -> String
+shapeOfValue (V1.Value m) =
+    let l = M.toList m
+    in printf "[%d: %s]" (length l) (intercalate "," (fmap (printf "%d" . length . M.toList . snd) l))
+
+analyseValue :: V1.Value -> IO ()
+analyseValue v = do
+  putStr $ shapeOfValue v
+  printf "\n"
+
+analyseTxInfoV1 :: V1.TxInfo -> IO ()
+analyseTxInfoV1 i = do
+  putStr "Fee:     "
+  analyseValue $ V1.txInfoFee i
+  putStr "Mint:    "
+  analyseValue $ V1.txInfoMint i
+  case V1.txInfoOutputs i of
+    [] -> putStrLn "No outputs"   -- This happens in 0000000046344292-*.event
+    l  -> do
+      putStr $ printf "Outputs %d " (length l)
+      putStrLn $ intercalate ", " (fmap (shapeOfValue . V1.txOutValue) l)
+
+analyseTxInfoV2 :: V2.TxInfo -> IO ()
```

Comment:
I've deduplicated it a bit, but there's still an annoying amount of almost-duplicated code in this file that's tricky to get rid of because many of the V1 and V2 types are different.